src/idl_gen_ts.cpp

```diff
@@ -219,68 +224,96 @@ class TsGenerator : public BaseGenerator {
+        // build path for root level vs child level
+        if (it.second.ns->components.size() > 1)
+          std::copy(it.second.ns->components.begin() + 1,
+                    it.second.ns->components.end(),
+                    std::back_inserter(rel_components));
+        else
+          std::copy(it.second.ns->components.begin(),
+                    it.second.ns->components.end(),
+                    std::back_inserter(rel_components));
+        auto base_file_name =
+            namer_.File(*(def.second), SkipFile::SuffixAndExtension);
+        auto base_name =
+            namer_.Directories(it.second.ns->components, SkipDir::OutputPath) +
+            base_file_name;
+        auto ts_file_path = base_name + ".ts";
+        auto base_name_rel = std::string("./");
+        base_name_rel +=
+            namer_.Directories(rel_components, SkipDir::OutputPath);
+        base_name_rel += base_file_name;
+        auto ts_file_path_rel = base_name_rel + ".ts";
+        auto type_name = def.first;
+        code += "export { " + type_name + " } from '";
+        code += base_name_rel + ".js';\n";
+        export_counter++;
       }
 
-      code += flat_file_;
-      const std::string filename =
-          GeneratedFileName(path_, file_name_, parser_.opts);
-      SaveFile(filename.c_str(), code, false);
-    } else {
-      for (auto it = imports_all_.begin(); it != imports_all_.end(); it++) {
-        code += it->second.export_statement + "\n";
+      // re-export child namespace(s) in parent
+      const auto child_ns_level = it.second.ns->components.size() + 1;
+      for (const auto &it2 : ns_defs_) {
+        if (it2.second.ns->components.size() != child_ns_level) continue;
+        auto ts_file_path = it2.second.path + ".ts";
+        code += "export * as " + it2.second.symbolic_name + " from './";
+        code += it2.second.path + ".js';\n";
+        export_counter++;
       }
-      const std::string path =
+
+      if (export_counter > 0) SaveFile(it.second.filepath.c_str(), code, false);
+    }
+  }
+
+  bool generateBundle() {
+    if (parser_.opts.ts_flat_files) {
+      std::string inputpath;
+      std::string symbolic_name = file_name_;
+      // TODO: unsure when/why this was sensible
+      /*if (parser_.current_namespace_->components.size() > 0) {
+        std::string path = namer_.Directories(*parser_.current_namespace_,
+                                              SkipDir::TrailingPathSeperator);
+        inputpath = path + ".ts";
+        symbolic_name = parser_.current_namespace_->components.back();
+      } else {
+        inputpath = path_ + file_name_ + ".ts";
+      }*/
+      inputpath = path_ + file_name_ + ".ts";
+      std::string bundlepath =
           GeneratedFileName(path_, file_name_, parser_.opts);
-      SaveFile(path.c_str(), code, false);
+      bundlepath = bundlepath.substr(0, bundlepath.size() - 3) + ".js";
+      std::string cmd = "esbuild";
+      cmd += " ";
+      cmd += inputpath;
```

Comment:
This inputpath that gets printed seems to be wrong, at least when just `--ts-flat-files` is used?

Reply:
What you do mean it seems wrong? I get fx `esbuild monster_test.ts --format=cjs --bundle --outfile=monster_test_generated.js --external:flatbuffers` as part of the runs in `TypeScriptTest.py` and is also what I then run at https://github.com/bjornharrtell/flatbuffers/blob/ns-entry-points/tests/ts/TypeScriptTest.py#L74 except I use cjs extension on the output file to make node accept it in the context of a module package.

Here is also an example of the process run entirely manually from shell at `/tests/ts`:
```bash
$ ../../flatc --ts --ts-flat-files -I ../include_test ../monster_test.fbs
Entry point monster_test.ts generated.
A single file bundle can be created using fx. esbuild with:
> esbuild monster_test.ts --format=cjs --bundle --outfile=monster_test_generated.js --external:flatbuffers
$ esbuild monster_test.ts --format=cjs --bundle --outfile=monster_test_generated.js --external:flatbuffers

  monster_test_generated.js  89.4kb

⚡ Done in 11ms
```

Reply:
I'm not quite sure what is happening, then. Maybe it's one of the other flags? When I run it in my bazel branch (https://github.com/jkuszmaul/flatbuffers/tree/ns-entry-points) I get:

```
$ bazelisk build //tests/ts:typescript_ts_fbs  
INFO: Analyzed target //tests/ts:typescript_ts_fbs (0 packages loaded, 0 targets configured).
INFO: Found 1 target...                                                
ERROR: /home/james/flatbuffers/tests/ts/BUILD.bazel:5:22: Generating flatbuffer files for typescript_ts_fbs_srcs: //tests/ts:typescript_ts_fbs_srcs failed: (Exit 1): bash failed: error executing command /bin/bash -c ... (remaining 1 argument skipped)
                                                                                                          
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
++ bazel-out/host/bin/ts/compile_flat_file.runfiles/com_github_google_flatbuffers/flatc -I ./ -I bazel-out/k8-fastbuild/bin -I bazel-out/k8-fastbuild/bin -I bazel-out/host/bin/ts/compile_flat_file.runfiles/com_github_google_flatbuffers --gen-object-api --gen-mutable --reflect-names --gen-name-strings --ts-flat-files --keep-prefix --filename-suffix _generated --ts -o bazel-out/k8-fastbuild/bin/tests/ts tests/ts/typescript
_keywords.fbs                                                   
+ FLATC_OUTPUT='Entry point bazel-out/k8-fastbuild/bin/tests/ts/typescript_keywords.ts generated.
A single file bundle can be created using fx. esbuild with:
> esbuild bazel-out/k8-fastbuild/bin/tests/ts/typescript_keywords.ts --format=cjs --bundle --outfile=bazel-out/k8-fastbuild/bin/tests/ts/typescript_keywords_generated.js --external:flatbuffers'
++ bazel-out/host/bin/ts/compile_flat_file.runfiles/com_github_google_flatbuffers/flatc -I ./ -I bazel-out/k8-fastbuild/bin -I bazel-out/k8-fastbuild/bin -I bazel-out/host/bin/ts/compile_flat_file.runfiles/com_github_google_flatbuffers --gen-object-api --gen-mutable --reflect-names --gen-name-strings --ts-flat-files --keep-prefix --filename-suffix _generated --ts -o bazel-out/k8-fastbuild/bin/tests/ts tests/ts/typescript
_keywords.fbs                                                    
++ grep 'Entry point.*generated'                           
++ grep -o 'bazel-out.*ts'
+ TS_FILE=bazel-out/k8-fastbuild/bin/tests/ts/typescript_keywords.ts
++ dirname bazel-out/k8-fastbuild/bin/tests/ts/typescript_keywords.ts
+ ls bazel-out/k8-fastbuild/bin/tests/ts    
foobar                         
foobar.ts           
reflection          
reflection.ts                                   
typescript               
typescript.ts                                                             
+ bazel-out/host/bin/ts/compile_flat_file.runfiles/npm/node_modules/esbuild/bin/esbuild bazel-out/k8-fastbuild/bin/tests/ts/typescript_keywords.ts --format=cjs --bundle --outfile=bazel-out/k8-fastbuild/bin/tests/ts/typescript_keywords_generated.cjs --external:flatbuffers
✘ [ERROR] Could not resolve "bazel-out/k8-fastbuild/bin/tests/ts/typescript_keywords.ts"
```

Reply:
@bjornharrtell Did you have any additional insight on this? I think this is the main thing preventing me from finishing bazel integration. I can probably figure it out if I spend some time on it.

Reply:
Sorry I don't quite understand what is blocking you or what I can do.

Reply:
@jkuszmaul can you try to explain what you are stuck with more plainly and assuming I don't understand bazel?

Reply:
I'll try to get you a reproduction case (outside of bazel) later today (california time).

Reply:
Okay, from the code in this PR if you run the following command in the root of the repo (assuming that `./flatc` is available):

```
$ rm -rf output_dir && mkdir output_dir && ./flatc -I ./ -I tests/ts/ --gen-object-api --gen-mutable --reflect-names --gen-name-strings --ts-flat-files --keep-prefix --filename-suffix _generated --ts -o output_dir tests/ts/typescript_keywords.fbs
Entry point output_dir/typescript_keywords.ts generated.
A single file bundle can be created using fx. esbuild with:
> esbuild output_dir/typescript_keywords.ts --format=cjs --bundle --outfile=output_dir/typescript_keywords_generated.js --external:flatbuffers
```

The referenced `output_dir/typescript_keywords.ts` does not actually exist.

Reply:
@jkuszmaul confirmed and I now see I've handled typescript_keywords.fbs differently in the test. The output hint is wrong, current code doesn't generate `typescript_keywords.ts` at all. The current output for `typescript_keywords.fbs` is three modules `typescript.ts`, `foobar.ts` and `reflection.ts` which are entry points for the namespaces in the included fbs.

I'm not exactly sure at this point what the correct output really should be, but an educated guess is that that a `typescript_keywords.ts` *should* be generated that re-exports the three mentioned modules. I'm currently a bit confused as for why `typescript_keywords.ts` isn't being generated.

Reply:
The reason `typescript_keywords.ts` is not being generated as opposed to `monster_test.ts` is because it represents the root namespace. The logic that determines the entry points to generate loops definitions and for `typescript_keywords.fbs` there is no definitions in the root namespace, only other namespaces. 

I have added the missing logic so that a root namespace entry point is always generated, i.e `typescript_keywords.ts` is generated.